Set error status when duplicate markers are in the same MarkerArray

sloretz opened the pull request:
This pull request adds an errors status to the `MarkerArray` display when a single `MarkerArray` message with duplicate markers is received. This can happen when the publisher author forgot to set `id` or `ns`. Assuming all the `Marker` messages have the same action (add/modify), it results in only the last marker being displayed.

![duplicate_marker_check](https://user-images.githubusercontent.com/4175662/187317881-c7320349-9cbb-4fd2-b0af-01ed154c70a1.png)


<details><summary>Script for checking</summary>

```python3
import copy

import rclpy
from visualization_msgs.msg import Marker, MarkerArray


def frame_markers():
    tpl = Marker()
    tpl.header.frame_id = 'map'
    tpl.type = Marker.CYLINDER
    tpl.scale.x = 0.25
    tpl.scale.y = 0.25
    tpl.scale.z = 0.5
    tpl.color.a = 1.0
    tpl.ns = 'my_pose'

    msg_x = copy.deepcopy(tpl)
    msg_x.color.r = 1.0
    msg_x.pose.position.x = -0.5;
    msg_x.id = 0

    msg_y = copy.deepcopy(tpl)
    msg_y.color.g = 1.0
    msg_y.pose.position.x = 0.0;
    msg_y.id = 1

    msg_z = copy.deepcopy(tpl)
    msg_z.color.b = 1.0
    msg_z.pose.position.x = 0.5;
    # msg_z.id = 2
    msg_z.id = 1   # Duplicate!
    return [msg_x, msg_y, msg_z]


def main():
    rclpy.init()

    node = rclpy.create_node("rviz_pub")
    marker_pub = node.create_publisher(MarkerArray, "via_marker_array", 10)

    while True:
        message = MarkerArray()
        message.markers = frame_markers()
        marker_pub.publish(message)


if __name__ == "__main__":
    main()
```
</details>

sloretz committed:
Set error status when duplicate markers are in the same MarkerArray

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

EricCousineau-TRI commented:
nice! one concern though is performance - there any simple performance tests that can be done to check that this processing doesn't blow compute budget for 30 FPS when using something like 1000 markers?

Also, is there perhaps a more efficient way to bookkeep? e.g. `std::set<PairType>`, vs. searching in `std::vector<>`?
also, any chance this can just be a warning, with no other behavior changes? e.g.
```cc
std::set<PairType> existing;
bool found_duplicates = false;
for (marker : markers) {
  addMessage(copy_of(marker));
  if (!found_duplicates) {
    PairType pair{marker.ns, marker.id};
    const bool is_new = existing.insert(pair).second;
    found_duplicates = !is_new;
  }
}
if (found_duplicates) {
  warn();
}
```

sloretz commented:
> any chance this can just be a warning, with no other behavior changes? e.g.

I don't think it changes behavior, at least that was my intent since I'd like to backport this.

EricCousineau-TRI commented:
gotcha, missed that. and yeah, your setup looks great, but still, perhaps good to use something with comparison time less than `O(N^2)` (for overall loop) just in case `marker.id` is not strictly increasing?

also, is necessary to have one loop with the full critical section, or might it be best to process message / check for duplicates before obtaining lock?

soz for asking dumb questions, just want to check

sloretz committed:
Add benchmark

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

More benchmarks

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
> is there perhaps a more efficient way to bookkeep? e.g. std::set<PairType>, vs. searching in std::vector<>?

46382a06fc48af18e203612e126a6a92b269a86c adds a temporary benchmark looking at how id's affect performance.

```
56: 2022-08-29T18:17:01-07:00                      
56: Running /home/osrf/ws/ros2/build/rviz_default_plugins/bm891
56: Run on (24 X 4950.19 MHz CPU s)
56: CPU Caches:
56:   L1 Data 32 KiB (x12)
56:   L1 Instruction 32 KiB (x12)
56:   L2 Unified 512 KiB (x12)
56:   L3 Unified 32768 KiB (x2)
56: Load Average: 2.42, 2.64, 2.63
56: ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
56: ***WARNING*** Library was built as DEBUG. Timings may be affected.
56: ---------------------------------------------------------------------------------
56: Benchmark                                       Time             CPU   Iterations
56: ---------------------------------------------------------------------------------
56: no_check_10                                  7703 ns         7702 ns        91065
56: no_check_100                                69445 ns        69436 ns        10029
56: no_check_1000                              683527 ns       683407 ns         1024
56: no_check_10000                            6919471 ns      6918240 ns           99
56: increasing_ids__vector_of_pairs__10          8011 ns         8010 ns        87359
56: increasing_ids__vector_of_pairs__100        70618 ns        70608 ns         9884
56: increasing_ids__vector_of_pairs__1000      692104 ns       691993 ns         1014
56: increasing_ids__vector_of_pairs__10000    7022189 ns      7020993 ns           98
56: decreasing_ids__vector_of_pairs__10          8612 ns         8611 ns        81219
56: decreasing_ids__vector_of_pairs__100       103880 ns       103866 ns         6773
56: decreasing_ids__vector_of_pairs__1000     3669271 ns      3668772 ns          191
56: decreasing_ids__vector_of_pairs__10000  302649791 ns    302603300 ns            2
56: increasing_ids__set_insertion__10           10192 ns        10191 ns        68646
56: increasing_ids__set_insertion__100         104252 ns       104239 ns         6709
56: increasing_ids__set_insertion__1000       1133052 ns      1132858 ns          617
56: increasing_ids__set_insertion__10000     12522444 ns     12520509 ns           56
```

The best case of the current method (called vector_of_pairs) adds just a few percent of overhead, but the worst case is about 43 times slower than not checking for duplicates at 10,000 elements. The implementation using an `std::set` performs worse than the vector_of_pairs method until 100 elements, and afterwords it does better. It seems to be 50% to 80% slower than not checking.

EricCousineau-TRI commented:
Thanks! Given that (10, 10000) markers can take:
- (7us, 7ms) without checks
- (8us, 30ms) (!!!) for checks w/ `std::vector<>`
- (10us, 12ms) for checks w/ `std::set<>`

I think `std::set<>` for checks makes the most sense, since you might take end up taking the whole 33.3ms budget for 30 FPS in processing markers if using `std::vector<>`.
WDYT?

EricCousineau-TRI commented:
(as an aside, mayhaps `std::unordered_set<>` could help the performance at cost of memory; but it may be marginal in this case - https://stackoverflow.com/a/1349883/7829525 )

sloretz committed:
Use std::set with id before ns

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
> I think std::set<> for checks makes the most sense, since you might take end up taking the whole 33.3ms budget for 30 FPS in processing markers if using std::vector<>.

I added some more benchmarks and switched to `std::set` with `id` now first in the pair type in 3662e20.

```
56: 2022-08-30T09:02:03-07:00
56: Running /home/osrf/ws/ros2/build/rviz_default_plugins/bm891
56: Run on (24 X 4950.19 MHz CPU s)
56: CPU Caches:
56:   L1 Data 32 KiB (x12)
56:   L1 Instruction 32 KiB (x12)
56:   L2 Unified 512 KiB (x12)
56:   L3 Unified 32768 KiB (x2)
56: Load Average: 3.25, 5.05, 3.74
56: ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
56: ***WARNING*** Library was built as DEBUG. Timings may be affected.
56: ----------------------------------------------------------------------------------------
56: Benchmark                                              Time             CPU   Iterations
56: ----------------------------------------------------------------------------------------
56: no_check_10                                         7594 ns         7593 ns        92133
56: no_check_100                                       66750 ns        66737 ns        10404
56: no_check_1000                                     652384 ns       651472 ns         1018
56: no_check_10000                                   6640213 ns      6627208 ns          103
56: increasing_ids__vector_of_pairs__10                 7890 ns         7888 ns        88519
56: increasing_ids__vector_of_pairs__100               69045 ns        69036 ns         9576
56: increasing_ids__vector_of_pairs__1000             669675 ns       669581 ns         1043
56: increasing_ids__vector_of_pairs__10000           6984555 ns      6983393 ns          100
56: decreasing_ids__vector_of_pairs__10                 8390 ns         8389 ns        80266
56: decreasing_ids__vector_of_pairs__100               98628 ns        98618 ns         7103
56: decreasing_ids__vector_of_pairs__1000            3421370 ns      3420919 ns          205
56: decreasing_ids__vector_of_pairs__10000         280956496 ns    280915720 ns            2
56: increasing_ids__set_insertion__10                  10226 ns        10225 ns        68416
56: increasing_ids__set_insertion__100                102784 ns       102769 ns         6771
56: increasing_ids__set_insertion__1000              1101864 ns      1101723 ns          634
56: increasing_ids__set_insertion__10000            12099845 ns     12097916 ns           58
56: decreasing_ids__set_insertion__10                  10138 ns        10137 ns        69105
56: decreasing_ids__set_insertion__100                 99622 ns        99606 ns         6971
56: decreasing_ids__set_insertion__1000              1083470 ns      1083285 ns          639
56: decreasing_ids__set_insertion__10000            11988269 ns     11986155 ns           57
56: increasing_ids__set_insertion_id_first__10          9983 ns         9982 ns        70204
56: increasing_ids__set_insertion_id_first__100        96213 ns        96202 ns         7279
56: increasing_ids__set_insertion_id_first__1000     1004277 ns      1004132 ns          697
56: increasing_ids__set_insertion_id_first__10000   10756909 ns     10755110 ns           65
56: decreasing_ids__set_insertion_id_first__10          9953 ns         9951 ns        69744
56: decreasing_ids__set_insertion_id_first__100        96422 ns        96412 ns         7265
56: decreasing_ids__set_insertion_id_first__1000     1011766 ns      1011640 ns          692
56: decreasing_ids__set_insertion_id_first__10000   10752093 ns     10750556 ns           65
```

> as an aside, mayhaps std::unordered_set<> could help the performance at cost of memory;

I didn't try this one because it requires implementing a hash function for the pair.

EricCousineau-TRI commented:
Yup, those results look like a convincing case for a reasonable trade-off between best- and worse-case using set + id-first, thanks!

EricCousineau-TRI commented on rviz_default_plugins/src/rviz_default_plugins/displays/marker/marker_common.cpp:
```diff
@@ -153,8 +154,32 @@ void MarkerCommon::addMessage(const visualization_msgs::msg::Marker::ConstShared
+  std::set<pair_type> unique_markers;
+  bool found_duplicate = false;
+
+  std::unique_lock<std::mutex> lock(queue_mutex_);
```
nit (another small performance question) Should you defer obtaining the lock until after you have checked for duplicates? 
Just in case holding for extra 4ms (in 10000 marker case) may bottleneck other operations.

e.g.
```cc
// First add messages for display.
{
  lock(queue_mutex_);
  for (marker : markers) {
    addMessage(...);
  }
}
// Scan markers for duplicates.
...
```

sloretz replied:
I did another benchmark (this time in Release mode). Locking every time has an extra 12% performance penalty, which may be worth it. I updated the implementation to use `addMessage()`.

<details><summary>Benchmarks locking every time</summary>

```
56: 2022-09-01T14:12:44-07:00
56: Running /home/osrf/ws/ros2/build/rviz_default_plugins/bm891
56: Run on (24 X 4950.19 MHz CPU s)
56: CPU Caches:
56:   L1 Data 32 KiB (x12)
56:   L1 Instruction 32 KiB (x12)
56:   L2 Unified 512 KiB (x12)
56:   L3 Unified 32768 KiB (x2)
56: Load Average: 1.47, 3.43, 2.03
56: ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
56: ----------------------------------------------------------------------------------------------------
56: Benchmark                                                          Time             CPU   Iterations
56: ----------------------------------------------------------------------------------------------------
56: no_check_10                                                      407 ns          407 ns      1721440
56: no_check_100                                                    4152 ns         4151 ns       168903
56: no_check_1000                                                  44429 ns        44423 ns        16155
56: no_check_10000                                                444279 ns       444222 ns         1576
56: increasing_ids__vector_of_pairs__10                              441 ns          441 ns      1590162
56: increasing_ids__vector_of_pairs__100                            4384 ns         4383 ns       159650
56: increasing_ids__vector_of_pairs__1000                          45767 ns        45760 ns        15281
56: increasing_ids__vector_of_pairs__10000                        454666 ns       454583 ns         1539
56: decreasing_ids__vector_of_pairs__10                              457 ns          457 ns      1532148
56: decreasing_ids__vector_of_pairs__100                            6176 ns         6175 ns       110264
56: decreasing_ids__vector_of_pairs__1000                         169046 ns       169025 ns         4144
56: decreasing_ids__vector_of_pairs__10000                      11280253 ns     11278468 ns           61
56: increasing_ids__set_insertion__10                                662 ns          662 ns      1058992
56: increasing_ids__set_insertion__100                              9493 ns         9491 ns        73724
56: increasing_ids__set_insertion__1000                           115436 ns       115415 ns         6060
56: increasing_ids__set_insertion__10000                         1265694 ns      1265513 ns          549
56: decreasing_ids__set_insertion__10                                667 ns          667 ns      1043784
56: decreasing_ids__set_insertion__100                              8455 ns         8454 ns        82235
56: decreasing_ids__set_insertion__1000                           106985 ns       106970 ns         6579
56: decreasing_ids__set_insertion__10000                         1202009 ns      1201840 ns          579
56: increasing_ids__set_insertion_id_first__10                       632 ns          632 ns      1107489
56: increasing_ids__set_insertion_id_first__100                     7880 ns         7879 ns        87671
56: increasing_ids__set_insertion_id_first__1000                   96081 ns        96069 ns         7296
56: increasing_ids__set_insertion_id_first__10000                1019056 ns      1018910 ns          680
56: decreasing_ids__set_insertion_id_first__10                       627 ns          626 ns      1109505
56: decreasing_ids__set_insertion_id_first__100                     7718 ns         7718 ns        91050
56: decreasing_ids__set_insertion_id_first__1000                   93514 ns        93501 ns         7493
56: decreasing_ids__set_insertion_id_first__10000                1004368 ns      1004208 ns          699
56: increasing_ids__set_insertion_id_first_always_lock__10           675 ns          675 ns      1035295
56: increasing_ids__set_insertion_id_first_always_lock__100         8326 ns         8325 ns        84129
56: increasing_ids__set_insertion_id_first_always_lock__1000      101719 ns       101706 ns         6912
56: increasing_ids__set_insertion_id_first_always_lock__10000    1076943 ns      1076798 ns          648
[Processing: rviz_default_plugins]                   
56: decreasing_ids__set_insertion_id_first_always_lock__10           702 ns          702 ns      1000924
56: decreasing_ids__set_insertion_id_first_always_lock__100         8117 ns         8115 ns        86392
56: decreasing_ids__set_insertion_id_first_always_lock__1000       98264 ns        98250 ns         7174
56: decreasing_ids__set_insertion_id_first_always_lock__10000    1054231 ns      1054092 ns          659
```
</details>

EricCousineau-TRI replied:
thanks!

sloretz committed:
Lock/Unlock for every message

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
CI (build: `--packages-above-and-dependencies rviz_default_plugins` test: `--packages-above rviz_default_plugins`)
* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17215)](http://ci.ros2.org/job/ci_linux/17215/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11758)](http://ci.ros2.org/job/ci_linux-aarch64/11758/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=17675)](http://ci.ros2.org/job/ci_windows/17675/)

EricCousineau-TRI reviewed (approved):
Approved from my perspective. Any chance there's a maintainer available to review?

jacobperron commented on rviz_default_plugins/src/rviz_default_plugins/displays/marker/marker_common.cpp:
```diff
@@ -153,9 +154,32 @@ void MarkerCommon::addMessage(const visualization_msgs::msg::Marker::ConstShared
+
   for (auto const & marker : array->markers) {
+    if (!found_duplicate) {
+      pair_type pair(marker.id, marker.ns);
+      found_duplicate = !unique_markers.insert(pair).second;
```
Consider using `emplace` if performance is a concern. I'm not sure if moves the needle on your benchmark though:

```suggestion
      found_duplicate = !unique_markers.emplace(marker.id, marker.ns).second;
```

sloretz replied:
I tried out in a benchmark, but I can't see a difference between `emplace` and `insert` within the noise.

<details><summary>One of the runs</summary>

```
56: 2022-09-02T13:58:15-07:00                      
56: Running /home/osrf/ws/ros2/build/rviz_default_plugins/bm891
56: Run on (24 X 4950.19 MHz CPU s)
56: CPU Caches:
56:   L1 Data 32 KiB (x12)
56:   L1 Instruction 32 KiB (x12)
56:   L2 Unified 512 KiB (x12)
56:   L3 Unified 32768 KiB (x2)
56: Load Average: 1.67, 2.49, 1.99
56: ***WARNING*** CPU scaling is enabled, the benchmark real time measurements may be noisy and will incur extra overhead.
56: ------------------------------------------------------------------------------------------------------------
56: Benchmark                                                                  Time             CPU   Iterations
56: ------------------------------------------------------------------------------------------------------------
56: no_check_10                                                              494 ns          494 ns      1390157
56: no_check_100                                                            5035 ns         5034 ns       139124
56: no_check_1000                                                          52140 ns        52134 ns        13403
56: no_check_10000                                                        522287 ns       522195 ns         1330
56: increasing_ids__vector_of_pairs__10                                      540 ns          540 ns      1293151
56: increasing_ids__vector_of_pairs__100                                    5126 ns         5124 ns       136431
56: increasing_ids__vector_of_pairs__1000                                  53451 ns        53444 ns        13273
56: increasing_ids__vector_of_pairs__10000                                526744 ns       526664 ns         1291
56: decreasing_ids__vector_of_pairs__10                                      552 ns          552 ns      1266562
56: decreasing_ids__vector_of_pairs__100                                    7073 ns         7072 ns        97768
56: decreasing_ids__vector_of_pairs__1000                                 187947 ns       187922 ns         3712
56: decreasing_ids__vector_of_pairs__10000                              11610110 ns     11608107 ns           60
56: increasing_ids__set_insertion__10                                        730 ns          730 ns       954514
56: increasing_ids__set_insertion__100                                     10345 ns        10341 ns        67748
56: increasing_ids__set_insertion__1000                                   121661 ns       121645 ns         5752
56: increasing_ids__set_insertion__10000                                 1360934 ns      1360740 ns          516
56: decreasing_ids__set_insertion__10                                        738 ns          738 ns       947607
56: decreasing_ids__set_insertion__100                                      9665 ns         9664 ns        72394
56: decreasing_ids__set_insertion__1000                                   120601 ns       120586 ns         5779
56: decreasing_ids__set_insertion__10000                                 1340419 ns      1340184 ns          526
56: increasing_ids__set_insertion_id_first__10                               708 ns          708 ns       987930
56: increasing_ids__set_insertion_id_first__100                             8838 ns         8837 ns        79525
56: increasing_ids__set_insertion_id_first__1000                          105766 ns       105751 ns         6617
56: increasing_ids__set_insertion_id_first__10000                        1141860 ns      1141685 ns          615
56: decreasing_ids__set_insertion_id_first__10                               724 ns          724 ns       967446
56: decreasing_ids__set_insertion_id_first__100                             8674 ns         8673 ns        80745
56: decreasing_ids__set_insertion_id_first__1000                          105557 ns       105542 ns         6615
56: decreasing_ids__set_insertion_id_first__10000                        1128819 ns      1128658 ns          631
56: increasing_ids__set_insertion_id_first_always_lock__10                   752 ns          752 ns       930742
56: increasing_ids__set_insertion_id_first_always_lock__100                 9304 ns         9303 ns        75296
56: increasing_ids__set_insertion_id_first_always_lock__1000              109530 ns       109515 ns         6393
56: increasing_ids__set_insertion_id_first_always_lock__10000            1175368 ns      1175185 ns          596
56: decreasing_ids__set_insertion_id_first_always_lock__10                   722 ns          722 ns       968217
56: decreasing_ids__set_insertion_id_first_always_lock__100                 9042 ns         9041 ns        77217
56: decreasing_ids__set_insertion_id_first_always_lock__1000              108964 ns       108949 ns         6395
56: decreasing_ids__set_insertion_id_first_always_lock__10000            1185452 ns      1185020 ns          598
56: increasing_ids__set_insertion_id_first_always_lock_emplace__10           757 ns          757 ns       921752                 
56: increasing_ids__set_insertion_id_first_always_lock_emplace__100         9250 ns         9249 ns        75718
56: increasing_ids__set_insertion_id_first_always_lock_emplace__1000      109513 ns       109499 ns         6318
56: increasing_ids__set_insertion_id_first_always_lock_emplace__10000    1190149 ns      1189968 ns          591
56: decreasing_ids__set_insertion_id_first_always_lock_emplace__10           759 ns          759 ns       920237
56: decreasing_ids__set_insertion_id_first_always_lock_emplace__100         9133 ns         9132 ns        76482
56: decreasing_ids__set_insertion_id_first_always_lock_emplace__1000      109207 ns       109191 ns         6436
56: decreasing_ids__set_insertion_id_first_always_lock_emplace__10000    1139710 ns      1139538 ns          613
```
</details>

jacobperron reviewed (approved):
LGTM, one comment below; take it or leave it.

wjwwood commented on rviz_default_plugins/src/rviz_default_plugins/displays/marker/marker_common.cpp:
```diff
@@ -153,9 +154,32 @@ void MarkerCommon::addMessage(const visualization_msgs::msg::Marker::ConstShared
+    display_->setStatusStd(
+      rviz_common::properties::StatusProperty::Error,
+      kDuplicateStatus,
+      "Multiple Markers in the same MarkerArray message had the same namespace and id");
```
Is this enough information for someone to debug this problem?

Maybe it should include how many are duplicates and/or which namespace/id are offending.

sloretz replied:
I made the error message include the first offending namespace and id in ba16715d08558469e3d2e9e4d8b70f294af710b7

wjwwood reviewed (approved):
Lgtm, aside from my comment about providing more debug info.

sloretz committed:
Output first offending namespace and id

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz committed:
Revert "Add benchmark"

This reverts commit 8aeea4c7f532c1650e41672d7a79970438ff8a98.

Signed-off-by: Shane Loretz <sloretz@osrfoundation.org>

sloretz commented:
CI Re-run (this time just FastRTPS because I don't think this change needs testing on multiple RMW)

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=17224)](http://ci.ros2.org/job/ci_linux/17224/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=11767)](http://ci.ros2.org/job/ci_linux-aarch64/11767/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=17686)](http://ci.ros2.org/job/ci_windows/17686/)

sloretz commented:
CMake warning on windows is unrelated to this PR

sloretz commented:
@mergifyio backport humble galactic foxy
